### PR TITLE
Feat/automate bench docker

### PR DESCRIPTION
**Pull request by heeckhau**
* Build and run the testbenches in Docker
* Plot the result (runtime vs latency) with the [Charming](https://github.com/yuankunzhang/charming) crate:
image

### CODE REVIEW

File: tlsn/benches/bin/plot.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        .expect("Usage: plot <path_to_csv_file>");
+
+    let mut rdr = csv::Reader::from_path(csv_file)?;
+
+    // Prepare data for plotting
+    let mut data = Vec::new();
+    for result in rdr.deserialize() {
+        let record: Metrics = result?;
+        // Summing upload_delay and download_delay for the x-axis value
+        let total_delay = record.upload_delay + record.download_delay;
```
Comment (yuroitaki) [inline tlsn/benches/bin/plot.rs:20]: nitpick: this line can go into `if record.name == "latency" {}` right?

File: tlsn/benches/bin/plot.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        if record.name == "latency" {
+            data.push(vec![total_delay as f32, record.runtime as f32]);
+        }
+    }
+
+    const TITLE: &str = "Runtime vs Latency";
+
+    // https://github.com/yuankunzhang/charming
+    let chart = Chart::new()
+        .x_axis(Axis::new())
```
Comment (yuroitaki) [inline tlsn/benches/bin/plot.rs:31]: can we label x and y axes for clarity too?

**Review by yuroitaki — COMMENTED** [review]
LGTM! just 2 nits haha

**Comment by heeckhau** [comment]
Thanks @yuroitaki , good feedback. I have improved the code.

I have also created a GitHub workflow to run this. It is currently triggered on `push` for testing, because manual triggers only work on the main branch.

Example run:
https://github.com/tlsnotary/tlsn/actions/runs/8390943304

File: tlsn/benches/bin/plot.rs [hunk 3]
```diff
@@ -13,25 +14,44 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
-        // Summing upload_delay and download_delay for the x-axis value
-        let total_delay = record.upload_delay + record.download_delay;
-        // Using runtime for the y-axis value
-        if record.name == "latency" {
-            data.push(vec![total_delay as f32, record.runtime as f32]);
-        }
-    }
+    let data: Vec<Vec<f32>> = rdr
+        .deserialize::<Metrics>()
+        .filter_map(Result::ok)
```
Comment (yuroitaki) [inline tlsn/benches/bin/plot.rs:19]: nice stuff, but we probably want to panic and exit if there is any data that can't be deserialized? Else the plot might be inaccurate?

**Review by yuroitaki — APPROVED** [review]
LGTM, great stuff!

**Comment by heeckhau** [comment]
@sinui0 the result now looks like this:
image

File: .github/workflows/bench.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+name: Run Benchmarks
+on:
+  # manual trigger
+  workflow_dispatch:
+  # push for testing
+  push:
```
[the remark this answers is not in the document]
Comment (heeckhau) [reply .github/workflows/bench.yml:6]: I agree. But manual triggers work on the main branch only.
It is currently enabled on push for testing only

**Pull request merged**